File: source/backend/src/main/java/ch/unisg/backend/controller/http/dto/response/SystemElementResponseDto.java
package ch.unisg.backend.controller.http.dto.response;

import ch.unisg.backend.controller.http.dto.response.ar.ConstraintResponseDto;
import ch.unisg.backend.core.domain.entities.classes.sos.SystemElement;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class SystemElementResponseDto {

    public static HashMap<String, Object> toJson(SystemElement systemElement) {

        HashMap<String, Object> systemElementDto = new HashMap<>();

        systemElementDto.put("id", systemElement.getId().toString());
        systemElementDto.put("title", systemElement.getTitle());

        List<HashMap<String, Object>> constraintListDto = new ArrayList<>();

        systemElement.getConstraintList().forEach(constraint -> {
            constraintListDto.add(ConstraintResponseDto.create(constraint));
        });

        systemElementDto.put("constraintList", constraintListDto);

        return systemElementDto;
    }
}

File: source/frontend/src/lib/dto/response/entity/SystemElementResponseDto.ts
import {
    ConstraintResponseDto,
    fromConstraintListResponseDto,
} from "$lib/dto/response/entity/ConstraintResponseDto";
import {SystemElement} from "$lib/domain/entity/sos/SystemElement.svelte";

export class SystemElementResponseDto {
    id : string | undefined;
    title : string | undefined;
    constraintList : ConstraintResponseDto[] = [] ;

    constructor() {}
}

export function fromSystemElementListResponseDto(systemElementResponseDto: SystemElementResponseDto[]) : SystemElement[] {
    return systemElementResponseDto.map(fromSystemElementResponseDto);
}

export function fromSystemElementResponseDto(systemElementResponseDto : SystemElementResponseDto) : SystemElement {
    const systemElement : SystemElement = SystemElement.create();

    systemElement.id = systemElementResponseDto.id;
    systemElement.title = systemElementResponseDto.title;

    systemElement.constraintList = fromConstraintListResponseDto(systemElementResponseDto.constraintList);

    return systemElement;
}